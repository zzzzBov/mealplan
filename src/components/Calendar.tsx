import React, { FC } from 'react';
import styled from 'styled-components';

export enum Month {
  January,
  February,
  March,
  April,
  May,
  June,
  July,
  August,
  September,
  October,
  November,
  December,
}

export enum DayOfWeek {
  Sunday,
  Monday,
  Tuesday,
  Wednesday,
  Thursday,
  Friday,
  Saturday,
}

const $Wrapper = styled.div``;

const $Table = styled.table``;

const $THead = styled.thead``;

const $HeaderRow = styled.tr``;

const $Header = styled.th``;

const $TBody = styled.tbody``;

const $WeekRow = styled.tr``;

const $WeekDay = styled.td``;

const $Date = styled.span``;

export const getWeekdays = (startOfWeek: DayOfWeek): string[] => {
  const weekdays = [];
  for (let i = 0; i < 7; i++) {
    const weekday = ((i + startOfWeek) % 7) as DayOfWeek;
    weekdays.push(DayOfWeek[weekday]);
  }
  return weekdays;
};

export const getWeeks = (month: Date, startOfWeek: DayOfWeek): Date[][] => {
  const year = month.getFullYear();
  const monthIndex = month.getMonth();
  const firstOfMonth = new Date(year, monthIndex, 1);
  const firstDayOfMonth = firstOfMonth.getDay();

  const dayDelta =
    startOfWeek - firstDayOfMonth - (startOfWeek > firstDayOfMonth ? 7 : 0);

  const startDate = new Date(year, monthIndex, 1 + dayDelta);

  const lastOfMonth = new Date(year, monthIndex + 1, 0);

  const weeks = [];
  for (let date = startDate; date <= lastOfMonth; ) {
    const week = [];
    for (let i = 0; i < 7; i++) {
      week.push(date);
      date = new Date(date.getFullYear(), date.getMonth(), date.getDate() + 1);
    }
    weeks.push(week);
  }

  return weeks;
};

interface ICalendarProps {
  month: Date;
  startOfWeek?: DayOfWeek;
}

export const Calendar: FC<ICalendarProps> = ({
  month,
  startOfWeek = DayOfWeek.Sunday,
}) => {
  const weekdays = getWeekdays(startOfWeek);
  const weeks = getWeeks(month, startOfWeek);

  return (
    <$Wrapper>
      <$Table>
        <$THead>
          <$HeaderRow>
            {weekdays.map((weekday, index) => (
              <$Header key={index}>{weekday}</$Header>
            ))}
          </$HeaderRow>
        </$THead>
        <$TBody>
          {weeks.map((week, weekIndex) => (
            <$WeekRow key={weekIndex}>
              {week.map((day, dayIndex) => (
                <$WeekDay key={dayIndex}>
                  <$Date>{day.getDate()}</$Date>
                </$WeekDay>
              ))}
            </$WeekRow>
          ))}
        </$TBody>
      </$Table>
    </$Wrapper>
  );
};
