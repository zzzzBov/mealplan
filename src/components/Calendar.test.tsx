import React from 'react';
import { render, screen } from '@testing-library/react';
import { Calendar, DayOfWeek, getWeekdays, getWeeks, Month } from './Calendar';

describe(`getWeekdays(startOfWeek)`, () => {
  it(`should be a function`, () => {
    expect(getWeekdays).toBeInstanceOf(Function);
  });

  ([
    [
      DayOfWeek.Sunday,
      [
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
      ],
    ],
    [
      DayOfWeek.Monday,
      [
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
        'Sunday',
      ],
    ],
    [
      DayOfWeek.Tuesday,
      [
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
        'Sunday',
        'Monday',
      ],
    ],
    [
      DayOfWeek.Wednesday,
      [
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
        'Sunday',
        'Monday',
        'Tuesday',
      ],
    ],
    [
      DayOfWeek.Thursday,
      [
        'Thursday',
        'Friday',
        'Saturday',
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
      ],
    ],
    [
      DayOfWeek.Friday,
      [
        'Friday',
        'Saturday',
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
      ],
    ],
    [
      DayOfWeek.Saturday,
      [
        'Saturday',
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
      ],
    ],
  ] as [DayOfWeek, string[]][]).forEach(([start, expected]) => {
    it(`should return the weekdays based on the start of the week`, () => {
      const result = getWeekdays(start);
      expect(result).toEqual(expected);
    });
  });
});

describe(`getWeeks(month, startOfWeek)`, () => {
  it(`should be a function`, () => {
    expect(getWeeks).toBeInstanceOf(Function);
  });

  ([
    [
      new Date(2020, Month.January),
      DayOfWeek.Sunday,
      [
        [
          new Date(2019, Month.December, 29),
          new Date(2019, Month.December, 30),
          new Date(2019, Month.December, 31),
          new Date(2020, Month.January, 1),
          new Date(2020, Month.January, 2),
          new Date(2020, Month.January, 3),
          new Date(2020, Month.January, 4),
        ],
        [
          new Date(2020, Month.January, 5),
          new Date(2020, Month.January, 6),
          new Date(2020, Month.January, 7),
          new Date(2020, Month.January, 8),
          new Date(2020, Month.January, 9),
          new Date(2020, Month.January, 10),
          new Date(2020, Month.January, 11),
        ],
        [
          new Date(2020, Month.January, 12),
          new Date(2020, Month.January, 13),
          new Date(2020, Month.January, 14),
          new Date(2020, Month.January, 15),
          new Date(2020, Month.January, 16),
          new Date(2020, Month.January, 17),
          new Date(2020, Month.January, 18),
        ],
        [
          new Date(2020, Month.January, 19),
          new Date(2020, Month.January, 20),
          new Date(2020, Month.January, 21),
          new Date(2020, Month.January, 22),
          new Date(2020, Month.January, 23),
          new Date(2020, Month.January, 24),
          new Date(2020, Month.January, 25),
        ],
        [
          new Date(2020, Month.January, 26),
          new Date(2020, Month.January, 27),
          new Date(2020, Month.January, 28),
          new Date(2020, Month.January, 29),
          new Date(2020, Month.January, 30),
          new Date(2020, Month.January, 31),
          new Date(2020, Month.February, 1),
        ],
      ],
    ],
    [
      new Date(2020, Month.January),
      DayOfWeek.Friday,
      [
        [
          new Date(2019, Month.December, 27),
          new Date(2019, Month.December, 28),
          new Date(2019, Month.December, 29),
          new Date(2019, Month.December, 30),
          new Date(2019, Month.December, 31),
          new Date(2020, Month.January, 1),
          new Date(2020, Month.January, 2),
        ],
        [
          new Date(2020, Month.January, 3),
          new Date(2020, Month.January, 4),
          new Date(2020, Month.January, 5),
          new Date(2020, Month.January, 6),
          new Date(2020, Month.January, 7),
          new Date(2020, Month.January, 8),
          new Date(2020, Month.January, 9),
        ],
        [
          new Date(2020, Month.January, 10),
          new Date(2020, Month.January, 11),
          new Date(2020, Month.January, 12),
          new Date(2020, Month.January, 13),
          new Date(2020, Month.January, 14),
          new Date(2020, Month.January, 15),
          new Date(2020, Month.January, 16),
        ],
        [
          new Date(2020, Month.January, 17),
          new Date(2020, Month.January, 18),
          new Date(2020, Month.January, 19),
          new Date(2020, Month.January, 20),
          new Date(2020, Month.January, 21),
          new Date(2020, Month.January, 22),
          new Date(2020, Month.January, 23),
        ],
        [
          new Date(2020, Month.January, 24),
          new Date(2020, Month.January, 25),
          new Date(2020, Month.January, 26),
          new Date(2020, Month.January, 27),
          new Date(2020, Month.January, 28),
          new Date(2020, Month.January, 29),
          new Date(2020, Month.January, 30),
        ],
        [
          new Date(2020, Month.January, 31),
          new Date(2020, Month.February, 1),
          new Date(2020, Month.February, 2),
          new Date(2020, Month.February, 3),
          new Date(2020, Month.February, 4),
          new Date(2020, Month.February, 5),
          new Date(2020, Month.February, 6),
        ],
      ],
    ],
  ] as [Date, DayOfWeek, Date[][]][]).forEach(([month, start, expected]) => {
    it(`should generate the week data for the specified month with the specified start date`, () => {
      const result = getWeeks(month, start);

      expect(result).toEqual(expected);
    });
  });
});

describe.skip(`<Calendar month={date} />`, () => {
  it(`should render a calendar for the specific month`, () => {
    const { container } = render(<Calendar month={new Date(2020, 0)} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <table>
          <thead>
            <tr>
              <th>Sunday</th>
              <th>Monday</th>
              <th>Tuesday</th>
              <th>Wednesday</th>
              <th>Thursday</th>
              <th>Friday</th>
              <th>Saturday</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td><span>29</span></td>
              <td><span>30</span></td>
              <td><span>31</span></td>
              <td><span>01</span></td>
              <td><span>02</span></td>
              <td><span>03</span></td>
              <td><span>04</span></td>
            </tr>
            <tr>
              <td><span>05</span></td>
              <td><span>06</span></td>
              <td><span>07</span></td>
              <td><span>08</span></td>
              <td><span>09</span></td>
              <td><span>10</span></td>
              <td><span>11</span></td>
            </tr>
            <tr>
              <td><span>12</span></td>
              <td><span>13</span></td>
              <td><span>14</span></td>
              <td><span>15</span></td>
              <td><span>16</span></td>
              <td><span>17</span></td>
              <td><span>18</span></td>
            </tr>
            <tr>
              <td><span>19</span></td>
              <td><span>20</span></td>
              <td><span>21</span></td>
              <td><span>22</span></td>
              <td><span>23</span></td>
              <td><span>24</span></td>
              <td><span>25</span></td>
            </tr>
            <tr>
              <td><span>26</span></td>
              <td><span>27</span></td>
              <td><span>28</span></td>
              <td><span>29</span></td>
              <td><span>30</span></td>
              <td><span>31</span></td>
              <td><span>01</span></td>
            </tr>
          </tbody>
        </table>
      </div>
    `);
  });
});
