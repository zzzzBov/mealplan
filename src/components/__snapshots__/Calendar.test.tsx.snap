// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar month={date} /> should render a calendar for the specific month 1`] = `
<div>
  <div
    class="sc-bdfBwQ"
  >
    <table
      class="sc-gsTCUz"
    >
      <thead
        class="sc-dlfnbm"
      >
        <tr
          class="sc-hKgILt"
        >
          <th
            class="sc-eCssSg"
          >
            Sunday
          </th>
          <th
            class="sc-eCssSg"
          >
            Monday
          </th>
          <th
            class="sc-eCssSg"
          >
            Tuesday
          </th>
          <th
            class="sc-eCssSg"
          >
            Wednesday
          </th>
          <th
            class="sc-eCssSg"
          >
            Thursday
          </th>
          <th
            class="sc-eCssSg"
          >
            Friday
          </th>
          <th
            class="sc-eCssSg"
          >
            Saturday
          </th>
        </tr>
      </thead>
      <tbody
        class="sc-jSgupP"
      >
        <tr
          class="sc-gKsewC"
        >
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              29
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              30
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              31
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              1
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              2
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              3
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              4
            </span>
          </td>
        </tr>
        <tr
          class="sc-gKsewC"
        >
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              5
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              6
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              7
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              8
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              9
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              10
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              11
            </span>
          </td>
        </tr>
        <tr
          class="sc-gKsewC"
        >
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              12
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              13
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              14
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              15
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              16
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              17
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              18
            </span>
          </td>
        </tr>
        <tr
          class="sc-gKsewC"
        >
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              19
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              20
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              21
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              22
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              23
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              24
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              25
            </span>
          </td>
        </tr>
        <tr
          class="sc-gKsewC"
        >
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              26
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              27
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              28
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              29
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              30
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              31
            </span>
          </td>
          <td
            class="sc-iBPRYJ"
          >
            <span
              class="sc-fubCfw"
            >
              1
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
